ffi/src/error.rs

```diff
@@ -0,0 +1,205 @@
+            Error::DeletionVector(_) => KernelError::DeletionVectorError,
+            Error::InvalidUrl(_) => KernelError::InvalidUrlError,
+            Error::MalformedJson(_) => KernelError::MalformedJsonError,
+            Error::MissingMetadata => KernelError::MissingMetadataError,
+            Error::MissingProtocol => KernelError::MissingProtocolError,
+            Error::InvalidProtocol(_) => KernelError::InvalidProtocolError,
+            Error::MissingMetadataAndProtocol => KernelError::MissingMetadataAndProtocolError,
+            Error::ParseError(..) => KernelError::ParseError,
+            Error::JoinFailure(_) => KernelError::JoinFailureError,
+            Error::Utf8Error(_) => KernelError::Utf8Error,
+            Error::ParseIntError(_) => KernelError::ParseIntError,
+            Error::InvalidColumnMappingMode(_) => KernelError::InvalidColumnMappingModeError,
+            Error::InvalidTableLocation(_) => KernelError::InvalidTableLocationError,
+            Error::InvalidDecimal(_) => KernelError::InvalidDecimalError,
+            Error::InvalidStructData(_) => KernelError::InvalidStructDataError,
+            Error::InternalError(_) => KernelError::InternalError,
+            Error::Backtraced {
+                source,
+                backtrace: _,
+            } => Self::from(*source),
+            Error::InvalidExpressionEvaluation(_) => KernelError::InvalidExpression,
+            Error::InvalidLogPath(_) => KernelError::InvalidLogPath,
+            Error::InvalidCommitInfo(_) => KernelError::InvalidCommitInfo,
+            Error::FileAlreadyExists(_) => KernelError::FileAlreadyExists,
+            Error::MissingCommitInfo => KernelError::MissingCommitInfo,
+            Error::Unsupported(_) => KernelError::UnsupportedError,
+            Error::ParseIntervalError(_) => KernelError::ParseIntervalError,
+            Error::ChangeDataFeedUnsupported(_) => KernelError::ChangeDataFeedUnsupported,
+        }
+    }
+}
+
+/// An error that can be returned to the engine. Engines that wish to associate additional
+/// information can define and use any type that is [pointer
+/// interconvertible](https://en.cppreference.com/w/cpp/language/static_cast#pointer-interconvertible)
+/// with this one -- e.g. by subclassing this struct or by embedding this struct as the first member
+/// of a [standard layout](https://en.cppreference.com/w/cpp/language/data_members#Standard-layout)
+/// class.
+#[repr(C)]
+pub struct EngineError {
+    pub(crate) etype: KernelError,
+}
+
+/// Semantics: Kernel will always immediately return the leaked engine error to the engine (if it
+/// allocated one at all), and engine is responsible for freeing it.
+#[repr(C)]
+pub enum ExternResult<T> {
+    Ok(T),
+    Err(*mut EngineError),
+}
+
+pub type AllocateErrorFn =
+    extern "C" fn(etype: KernelError, msg: KernelStringSlice) -> *mut EngineError;
+
+// NOTE: We can't "just" impl From<DeltaResult<T>> because we require an error allocator.
```

Comment:
This comment seems misplaced? Maybe it belongs with the `IntoExternResult` trait below?

Reply:
seems reasonable - moved!